File: 线性表/栈及其应用/SequentialStack.java
public class SequentialStack<T> {
    public static final int DEFAULT_CAPACITY = 100;
    private T[] baseArray = (T[]) new Object[0];
    private int base = 0;
    private int top = base;
    private int capacity;

    public SequentialStack() {
        this(DEFAULT_CAPACITY);
    }

    public SequentialStack(int capacity) {
        this.capacity = capacity;
        ensureCapacity(capacity);
    }

    public boolean isEmpty() {
        return top - base == 0;
    }

    public int size() {
        return top - base;
    }

    private void ensureCapacity(int newSize) {
        if (newSize < top - base)
            return;

        T[] newArray = (T[]) new Object[newSize];
        for (int i = base; i < top; ++i)
            newArray[i - base] = baseArray[i];
        top = top - base;
        base = 0;
        baseArray = newArray;
    }

    public void trimToSize() {
        ensureCapacity(size());
    }

    public void clear() {
        top = base;
        ensureCapacity(capacity);
    }

    public void push(T data) {
        if (top == baseArray.length)
            ensureCapacity(baseArray.length + capacity);
        baseArray[top++] = data;
    }

    public T pop() {
        if (isEmpty())
            throw new RuntimeException("empty stack");
        return baseArray[--top];
    }

    public T getTop() {
        if (isEmpty())
            throw new RuntimeException("empty stack");
        return baseArray[top - 1];
    }

    public void println() {
        for (int i = base; i < top; ++i) 
            System.out.println(baseArray[i]);
    }

    public static int[] conversion(int n, int base) {
        SequentialStack<Integer> stack = new SequentialStack<Integer>();
        while (n != 0) {
            stack.push(n % base);
            n = n / base;
        }
        int size = stack.size();
        int[] result = new int[size];
        for (int i = 0; i < size; ++i)
            result[i] = stack.pop();
        return result;
    }

    public static boolean isPair(int[] array) {
        SequentialStack<Integer> stack = new SequentialStack<Integer>();
        for (int ele: array) {
            if (ele == 0)
                continue;
            if (ele < 0)
                stack.push(ele);
            else {
                if (stack.size() == 0 || (stack.pop() + ele != 0))
                    return false;
            }
        }
        return stack.size() == 0;
    }
}

File: 线性表/栈及其应用/sequential_stack.py
# coding: utf8


class Stack(object):
    def __init__(self):
        self._array_list = []
        self._base = self._top = 0

    def push(self, element):
        self._array_list.append(element)
        self._top = self._top + 1

    def is_empty(self):
        return self._base == self._top

    def size(self):
        return self._top - self._base

    def pop(self):
        if self.is_empty():
            raise RuntimeError("stack is empty")
        element = self._array_list.pop()
        self._top = self._top - 1
        return element

    def peek(self):
        if self.is_empty():
            raise RuntimeError("stack is empty")
        return self._array_list[-1]


def is_pair(string):
    lefts = {"{", "[", "("}
    matches = {"{": "}", "[": "]", "(": ")"}

    stack = Stack()
    for char in string:
        if char in lefts:
            # 遇到左括号，压栈
            stack.push(char)
        else:
            # 遇到右括号弹出栈顶元素，并判断是否匹配
            if stack.size() == 0:
                return False
            left = stack.pop()
            if char != matches[left]:
                return False

    # 最后检查栈是否为空
    return stack.size() == 0


def conversion(n, base):
    result = ["+"]
    if n < 0:
        result[0] = "-"
        n = -1 * n

    stack = Stack()
    while n != 0:
        stack.push(n % base)
        n = n / base

    while stack.size() > 0:
        result.append(stack.pop())

    return result


if __name__ == "__main__":
    print(is_pair("[[]]()"))
    print(is_pair("[[()()]"))
    print(conversion(35, 18))
